In [1]:
import pandas as pd

In [2]:
import numpy as np

In [3]:
import warnings
warnings.filterwarnings("ignore")

In [4]:
pd.set_option('display.max_columns', 30)
pd.set_option('display.width', 200)
pd.set_option('display.max_colwidth', 100)

In [5]:
data_path = 'C:/Users/Hugo/Documents/BTS - MDSD/Big Data/data/DF-Data_twitter_vine/twitter-vine-clean.tsv.bz2'

rawdf = pd.read_csv('C:/Users/Hugo/Documents/BTS - MDSD/Big Data/data/DF-Data_twitter_vine/twitter-vine-clean.tsv.bz2', 
                    sep='\t', compression='bz2')

In [6]:
print('Number of rows and columns:')
print(rawdf.shape)

Number of rows and columns:
(2129755, 22)


In [7]:
print('See the top 10 lines:')
rawdf.head(10)

See the top 10 lines:


created_at            tweet_id                                                                                                 text  \
0  2013-01-28T13:41:54+00:00  295889441355071488                    BBC News - Twitter launches Vine micro-video sharing service http://t.co/MykDuEgx   
1  2013-01-28T13:41:57+00:00  295889452121866241                               “Vine” อลวน โป๊ระบาด-เฟซบุ๊กสั่งแบน-ทวิตเตอร์เซ็ง http://t.co/mIo9PO6t   
2  2013-01-28T13:41:58+00:00  295889454705565696  Really excited for the explosion of 'What does Vine mean for _______?' pieces, which will dodge ...   
3  2013-01-28T13:41:58+00:00  295889458639802368  It's fun to think about the potential that exists between #vine and those who have DIY #Etsy bus...   
4  2013-01-28T13:42:01+00:00  295889470333517824                              Just Vined Lets You See the Last 20 Videos on Vine http://t.co/5tqQtM92   
5  2013-01-28T13:42:01+00:00  295889470836850688                                   Анимированные GIF-ки и короткие видео в новом сервисе Twitter Vine   
6  2013-01-28T13:42:03+00:00  295889475635134464  Epoke la vine tendu pa attann moun vine tann bras a yo, pass life la tepu sa u baw fèy rèprany o...   
7  2013-01-28T13:42:03+00:00  295889478600495104                                                            Vine a perder mi tiempo, sin comentarios.   
8  2013-01-28T13:42:03+00:00  295889478831198208                                                            http://t.co/G5vwi7u3 http://t.co/Sx2tn2wV   
9  2013-01-28T13:42:04+00:00  295889481045782528             Brands Are Already Experimenting With Twitter’s Vine http://t.co/kd1lGfeY via @comunitee   

                                                                                 source favorited retweeted possibly_sensitive lang   geo coordinates  \
0              <a href="http://twitter.com/tweetbutton" rel="nofollow">Tweet Button</a>     False     False              False   en  None        None   
1                 <a href="http://www.facebook.com/twitter" rel="nofollow">Facebook</a>     False     False              False   th  None        None   
2                                                                                   web     False     False                NaN   en  None        None   
3                                                                                   web     False     False                NaN   en  None        None   
4                                                                                   web     False     False              False   en  None        None   
5                                                                                   web     False     False                NaN   ru  None        None   
6                                                                                   web     False     False                NaN   fr  None        None   
7  <a href="http://twitter.com/download/android" rel="nofollow">Twitter for Android</a>     False     False                NaN   es  None        None   
8                 <a href="http://www.facebook.com/twitter" rel="nofollow">Facebook</a>     False     False              False  und  None        None   
9                      <a href="http://www.comunitee.com/" rel="nofollow">Comunitee</a>     False     False              False   en  None        None   

                                                                         hashtags                                                                                                 urls  \
0                                                                              []  [{'url': 'http://t.co/MykDuEgx', 'indices': [61, 81], 'display_url': 'bbc.in/148Grra', 'expanded...   
1                                                                              []  [{'url': 'http://t.co/mIo9PO6t', 'indices': [50, 70], 'display_url': 'manager.co.th/Cyberbiz/Vie...   
2                                                                              []           

### Question 1 - Total number of tweets?

In [8]:
print('Total number of tweets:')

rawdf.tweet_id.count()

Total number of tweets:


2129755

### Question 2 - Total number of unique users?

In [9]:
print('Total number of unique users:')

rawdf.user_id.nunique()

Total number of unique users:


918911

### Question 3 - How many tweets contains more than one URL?

In [10]:
print('Number of tweets containing at least one URL:')

rawdf.text.str.contains('http').sum()

Number of tweets containing at least one URL:


1549768

In [11]:
print('How many tweets contains more than one URL?')

urldf = pd.DataFrame(rawdf.text.str.split('http',2).tolist(),
                     columns = ['body','http1','http2'])

urldf.head()

How many tweets contains more than one URL?


body             http1 http2
0                                        BBC News - Twitter launches Vine micro-video sharing service   ://t.co/MykDuEgx  None
1                                                   “Vine” อลวน โป๊ระบาด-เฟซบุ๊กสั่งแบน-ทวิตเตอร์เซ็ง   ://t.co/mIo9PO6t  None
2  Really excited for the explosion of 'What does Vine mean for _______?' pieces, which will dodge ...              None  None
3  It's fun to think about the potential that exists between #vine and those who have DIY #Etsy bus...              None  None
4                                                  Just Vined Lets You See the Last 20 Videos on Vine   ://t.co/5tqQtM92  None

In [12]:
print('Total number of None values in http2 column:')

urldf.http2.isnull().sum() #get this number and take it away from the total number of tweets (2129755)

Total number of None values in http2 column:


2109856

We can see that **2109856** tweets don't contain a second URL in their text.  If we take this number away from the total number of tweets, 2129755, we can see tht **19,899** tweets contain more than 1 URL.

A more elegant way to solve this question was as follows:

In [13]:
sum(rawdf.text.str.count('http')>1)

19899

### Question 4 - How many tweets are geotagged (have latitude and longitude)?

In [14]:
print('Total number of tweets:')

rawdf.tweet_id.count()

Total number of tweets:


2129755

In [15]:
print('Tweets grouped by geotag:')

geodf = pd.DataFrame(rawdf.groupby(['geo']).apply(len))
geodf.sort(columns=[0], ascending=False).head(5)

Tweets grouped by geotag:


0
geo                                                                    
None                                                            2114183
{'type': 'Point', 'coordinates': [0.0, 0.0]}                         74
{'type': 'Point', 'coordinates': [34.10162416, -118.32664847]}       64
{'type': 'Point', 'coordinates': [53.800651, -4.064941]}             38
{'type': 'Point', 'coordinates': [54.0633, -2.88424]}                32

We can see that of 2129755 tweets, 2114183 have no geotag.  This leaves **15,572** tweets that are geotagged with both  both a longitude and latitude specified for their coordinates. 
*74* tweets are specified as having coordinates at [0.0, 0.0].

### Question 5 - How many tweets are original (e.g., not retweets)?

In [16]:
retweetdf = pd.DataFrame(rawdf.groupby(['retweeted']).apply(len))
retweetdf.sort(columns=[0], ascending=False)

0
retweeted         
False      2129597
True             5

We can see that just **5** of the tweets are retweets.
**2129597** tweets are original.

### Question 6 - How many tweets for each day? In average?

In [15]:
print('Number of tweets on each day:')

daydf = pd.DataFrame(rawdf.created_at.str.split('T',1).tolist(),
                     columns = ['day','created_at'])
daysdf = pd.DataFrame(daydf.groupby(['day']).apply(len))
daysdf.head(50)

Number of tweets on each day:


0
day              
2013-01-28  69123
2013-01-29  98501
2013-01-30  75435
2013-01-31  71642
2013-02-01  75419
2013-02-02  72197
2013-02-03  46374
2013-02-04  34824
2013-02-05  62784
2013-02-06  79157
2013-02-07  72738
2013-02-08  80017
2013-02-09  87639
2013-02-10  79522
2013-02-11  64506
2013-02-12  69265
2013-02-13  75196
2013-02-14  69675
2013-02-15  72621
2013-02-16  78666
2013-02-17  74956
2013-02-18  71359
2013-02-19  70309
2013-02-20  72000
2013-02-21  77571
2013-02-22  80345
2013-02-23  89468
2013-02-24  88286
2013-02-25  70160

In [21]:
print('Number of tweets on each day in ascending order:')

daydf = pd.DataFrame(rawdf.created_at.str.split('T',1).tolist(),
                     columns = ['day','created_at'])
daysdf = pd.DataFrame(daydf.groupby(['day']).apply(len))
daysdf.sort(columns=[0], ascending=False)

Number of tweets on each day in ascending order:


0
day              
2013-01-29  98501
2013-02-23  89468
2013-02-24  88286
2013-02-09  87639
2013-02-22  80345
2013-02-08  80017
2013-02-10  79522
2013-02-06  79157
2013-02-16  78666
2013-02-21  77571
2013-01-30  75435
2013-02-01  75419
2013-02-13  75196
2013-02-17  74956
2013-02-07  72738
2013-02-15  72621
2013-02-02  72197
2013-02-20  72000
2013-01-31  71642
2013-02-18  71359
2013-02-19  70309
2013-02-25  70160
2013-02-14  69675
2013-02-12  69265
2013-01-28  69123
2013-02-11  64506
2013-02-05  62784
2013-02-03  46374
2013-02-04  34824

In [22]:
daysdf.mean(0).round(2)

0    73439.83
dtype: float64

The mean average of the number of tweets on a given day is **73439.83**.

### Question 7 - Day of the week with highest volume of traffic?

In [35]:
raw2df = pd.DataFrame({'day':['1-28-2013','1-29-2013','1-30-2013','1-31-2013','2-1-2013','2-2-2013','2-3-2013','2-4-2013','2-5-2013','2-6-2013','2-7-2013','2-8-2013','2-9-2013','2-10-2013','2-11-2013','2-12-2013','2-13-2013','2-14-2013','2-15-2013','2-16-2013','2-17-2013','2-18-2013','2-19-2013','2-20-2013','2-21-2013','2-22-2013','2-23-2013','2-24-2013','2-25-2013']})
                 
raw2df['day'] = pd.to_datetime(raw2df['day'])
raw2df['day_of_week'] = raw2df['day'].dt.dayofweek

raw2df.head(30)

day  day_of_week
0  2013-01-28            0
1  2013-01-29            1
2  2013-01-30            2
3  2013-01-31            3
4  2013-02-01            4
5  2013-02-02            5
6  2013-02-03            6
7  2013-02-04            0
8  2013-02-05            1
9  2013-02-06            2
10 2013-02-07            3
11 2013-02-08            4
12 2013-02-09            5
13 2013-02-10            6
14 2013-02-11            0
15 2013-02-12            1
16 2013-02-13            2
17 2013-02-14            3
18 2013-02-15            4
19 2013-02-16            5
20 2013-02-17            6
21 2013-02-18            0
22 2013-02-19            1
23 2013-02-20            2
24 2013-02-21            3
25 2013-02-22            4
26 2013-02-23            5
27 2013-02-24            6
28 2013-02-25            0

In [35]:
daysdf.head(2)

0
day              
2013-01-28  69123
2013-01-29  98501

In [36]:
#pd.merge(rawdf, daysdf, on='day', how='outer')

raw2df = pd.merge(raw2df, daysdf, on=raw2df.day, how='outer')

raw2df.head(30)

day  day_of_week      0
0  2013-01-28            0  69123
1  2013-01-29            1  98501
2  2013-01-30            2  75435
3  2013-01-31            3  71642
4  2013-02-01            4  75419
5  2013-02-02            5  72197
6  2013-02-03            6  46374
7  2013-02-04            0  34824
8  2013-02-05            1  62784
9  2013-02-06            2  79157
10 2013-02-07            3  72738
11 2013-02-08            4  80017
12 2013-02-09            5  87639
13 2013-02-10            6  79522
14 2013-02-11            0  64506
15 2013-02-12            1  69265
16 2013-02-13            2  75196
17 2013-02-14            3  69675
18 2013-02-15            4  72621
19 2013-02-16            5  78666
20 2013-02-17            6  74956
21 2013-02-18            0  71359
22 2013-02-19            1  70309
23 2013-02-20            2  72000
24 2013-02-21            3  77571
25 2013-02-22            4  80345
26 2013-02-23            5  89468
27 2013-02-24            6  88286
28 2013-02-25            0  70160

In [40]:
dowdf = pd.DataFrame(raw2df.groupby(['day_of_week']).apply(sum))
#dowdf.sort(columns=[0], ascending=False).head(10)
dowdf.head(10)

day_of_week       0
day_of_week                     
0                      0  309972
1                      4  300859
2                      8  301788
3                     12  291626
4                     16  308402
5                     20  327970
6                     24  289138

We can see that the day of the week with the most amount of traffic is day 5 in the day of the week index, **Saturday**, with **327970** tweets.

### Question 8 - Which is the most popular user_location?

In [40]:
print('Question 8 - Which is the most popular user_location?')

locdf = pd.DataFrame(rawdf.groupby(['user_location']).apply(len))
locdf.sort(columns=[0], ascending=False).head()

Question 8 - Which is the most popular user_location?


0
user_location         
London           21250
New York, NY     12095
Los Angeles      11237
New York         10892
Los Angeles, CA   9456

We can see that the most common user location in our dataset is London with **21250** tweets

### Question 9 - Which are the most popular hashtags?

In [41]:
hashdf = pd.DataFrame(rawdf.hashtags.str.split('text',3).tolist(),columns = ['body','hashtag1','hashtag2','hashtag3'])

hashdf.head()

AttributeError: 'DataFrame' object has no attribute 'hashtags'